sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala

```diff
@@ -111,12 +112,18 @@ case class BroadcastExchangeExec(
           SQLMetrics.postDriverMetricUpdates(sparkContext, executionId, metrics.values.toSeq)
           broadcasted
         } catch {
+          // SPARK-24294: To bypass scala bug: https://github.com/scala/bug/issues/9554, we throw
+          // SparkFatalException, which is a subclass of Exception. ThreadUtils.awaitResult
+          // will catch this exception and re-throw the wrapped fatal throwable.
           case oe: OutOfMemoryError =>
```

Comment:
not related to this PR, but I'm a little worried about catching OOM here. Spark has `SparkOutOfMemoryError`, and it seems more reasonable to catch `SparkOutOfMemoryError`. This can be fixed in another PR.

Reply:
Agreed; let's fix that separately.

Reply:
To be clear, @cloud-fan,  do you mean that, ideally  during `relationFuture`, all the OOM error thrown of type SparkOutOfMemoryError ? (SparkOutOfMemoryError subclass of OutOfMemoryError)

Reply:
yea, I think Spark will only throw `SparkOutOfMemoryError` and we should only extend the error message for `SparkOutOfMemoryError`.

Reply:
Sure, I'm glad take this and resolve in another pr.